src/plugins/fs/udf.c

```diff
@@ -0,0 +1,503 @@
+    DEPS_UDFLABEL_MASK,     /* set-uuid */
+};
+
+#define UNUSED __attribute__((unused))
+
+/**
+ * bd_fs_udf_is_tech_avail:
+ * @tech: the queried tech
+ * @mode: a bit mask of queried modes of operation (#BDFSTechMode) for @tech
+ * @error: (out): place to store error (details about why the @tech-@mode combination is not available)
+ *
+ * Returns: whether the @tech-@mode combination is available -- supported by the
+ *          plugin implementation and having all the runtime dependencies available
+ */
+gboolean __attribute__ ((visibility ("hidden")))
+bd_fs_udf_is_tech_avail (BDFSTech tech UNUSED, guint64 mode, GError **error) {
+    guint32 required = 0;
+    guint i = 0;
+
+    if (mode & BD_FS_TECH_MODE_CHECK || mode & BD_FS_TECH_MODE_REPAIR) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_TECH_UNAVAIL,
+                     "UDF doesn't support checking and reparing.");
+        return FALSE;
+    } else if (mode & BD_FS_TECH_MODE_RESIZE) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_TECH_UNAVAIL,
+                     "UDF currently doesn't support resizing.");
+        return FALSE;
+    }
+
+    for (i = 0; i <= BD_FS_MODE_LAST; i++)
+        if (mode & (1 << i))
+            required |= fs_mode_util[i];
+
+    return check_deps (&avail_deps, required, deps, DEPS_LAST, &deps_check_lock, error);
+}
+
+/**
+ * bd_fs_udf_info_copy: (skip)
+ *
+ * Creates a new copy of @data.
+ */
+BDFSUdfInfo* bd_fs_udf_info_copy (BDFSUdfInfo *data) {
+    if (data == NULL)
+        return NULL;
+
+    BDFSUdfInfo *ret = g_new0 (BDFSUdfInfo, 1);
+
+    ret->label = g_strdup (data->label);
+    ret->uuid = g_strdup (data->uuid);
+    ret->revision = g_strdup (data->revision);
+    ret->lvid = g_strdup (data->lvid);
+    ret->vid = g_strdup (data->vid);
+    ret->block_size = data->block_size;
+    ret->block_count = data->block_count;
+    ret->free_blocks = data->free_blocks;
+
+    return ret;
+}
+
+/**
+ * bd_fs_udf_info_free: (skip)
+ *
+ * Frees @data.
+ */
+void bd_fs_udf_info_free (BDFSUdfInfo *data) {
+    if (data == NULL)
+        return;
+
+    g_free (data->label);
+    g_free (data->uuid);
+    g_free (data->revision);
+    g_free (data->lvid);
+    g_free (data->vid);
+    g_free (data);
+}
+
+BDExtraArg __attribute__ ((visibility ("hidden")))
+**bd_fs_udf_mkfs_options (BDFSMkfsOptions *options, const BDExtraArg **extra) {
+    GPtrArray *options_array = g_ptr_array_new ();
+    const BDExtraArg **extra_p = NULL;
+    g_autofree gchar *vid = NULL;
+
+    if (options->label) {
+        /* there are different limits for lvid and vid and we need to set both so lets just
+           truncate vid (GParted does the same thing) */
+        if (g_str_is_ascii (options->label) && strlen (options->label) > 30)
+            vid = g_strndup (options->label, 30);
+        else if (g_utf8_validate (options->label, -1, NULL) && g_utf8_strlen (options->label, -1) > 15)
+            vid = g_utf8_substring (options->label, 0, 15);
```

Comment:
Hello! You have there incorrect checks in validation. See restrictions in: https://github.com/storaged-project/udisks/issues/381